Support Alibaba-NLP/gte-Qwen2-7B-instruct embedding Model

zhaochenyang20 opened the pull request:
## Motivation

Current SGLang only supports the e5-mistral embedding model. I added Alibaba-NLP/gte-Qwen2-7B-instruct model in this PR. 

Also, previously SGLang determines a model as an embedding model through its `hf_config.architectures`. But gte model has the same architecture as CausalLM. So I added a new parameter in the `server_args` and changed the forward function of `Qwen2ForCausalLM`.



## Modifications

1. Changed the forward function of `Qwen2ForCausalLM`.
2. Added a new parameter `is_embedding` in `server_args`.
3. Some related changes.
4. Added unit tests for gte models. (both in the generation and embedding tests. I used rouge-L score in the generation tests)
5. Changed readme.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

Ying1123 commented on python/sglang/srt/managers/tp_worker.py:
```diff
@@ -844,6 +845,7 @@ def run_tp_server(
 ):
     """Run a tensor parallel server."""
     try:
+
```
```suggestion
```

Ying1123 commented on python/sglang/srt/managers/tp_worker.py:
```diff
@@ -871,6 +873,7 @@ def launch_tp_servers(
     """Launch multiple tensor parallel servers."""
     procs = []
     for i in tp_rank_range:
+
```
```suggestion
```

Ying1123 commented on python/sglang/srt/model_executor/model_runner.py:
```diff
@@ -196,7 +196,7 @@ def load_model(self):
         )
         self.is_generation = is_generation_model(
-            self.model_config.hf_config.architectures
+            self.model_config.hf_config.architectures, self.server_args
```
```suggestion
            self.model_config.hf_config.architectures, self.server_args.is_embedding
```

Ying1123 commented on python/sglang/srt/model_executor/model_runner.py:
```diff
@@ -498,9 +498,18 @@ def forward_extend(self, batch: ScheduleBatch):
-        return self.model.forward(
-            batch.input_ids, input_metadata.positions, input_metadata
-        )
+        if self.is_generation == True:
```
```suggestion
        if self.is_generation:
```

Ying1123 commented on python/sglang/srt/models/qwen2.py:
```diff
@@ -283,11 +285,15 @@ def forward(
+        else:
+            return self.logits_processor(
+                input_ids, hidden_states, self.lm_head.weight, input_metadata
+            )
```
Change the order of two branches. We always choose the generation branch as the first one.

Ying1123 commented on python/sglang/srt/server.py:
```diff
@@ -307,6 +307,7 @@ def launch_server(
                 (server_args.node_rank + 1) * tp_size_local,
             )
         )
+        #! 这里 launch 了 tp server
```
```suggestion
```

Ying1123 commented on python/sglang/srt/server.py:
```diff
@@ -500,6 +503,7 @@ def __init__(
     ):
         """See the arguments in server_args.py::ServerArgs"""
         self.server_args = ServerArgs(*args, log_level=log_level, **kwargs)
+        assert self.server_args.is_embedding is not None, "is_embedding must be set."
```
```suggestion
```

Ying1123 commented on test/srt/models/test_generation_models.py:
Add `Alibaba-NLP/gte-Qwen2-1.5B-instruct` to this test.

zhaochenyang20 commented:
@Ying1123 I added gte in the generation model test. Note that I changed the prefill tolerance accordingly and added the rouge-l metric instead of assert output_strs exactly the same.

Ying1123 committed:
add support for gte embedding

Ying1123 committed:
skip last prompt for gte (passed on h100 but not our a100 CI machine)

hnyls2002 commented on test/srt/models/test_generation_models.py:
```diff
@@ -56,17 +94,34 @@ def assert_close_prefill_logits_and_output_strs(
+        import multiprocessing as mp
+
+        try:
+            mp.set_start_method("spawn")
```
Why would this be needed?

llmforever commented:
different result compare with orginal hf back end,why?

prompt = “hello world”
sglang:
import openai
client = openai.Client(
    base_url="http://localhost:30000/v1", api_key="EMPTY")

response = client.embeddings.create(
    model="default",
    input=prompt ,
)

transformer：
tokenizer = AutoTokenizer.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)
model = AutoModel.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)

max_length = 8192

# Tokenize the input texts
batch_dict = tokenizer(prompt, max_length=max_length, padding=True, truncation=True, return_tensors='pt')
outputs = model(**batch_dict)
embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

# normalize embeddings
embeddings = F.normalize(embeddings, p=2, dim=1)

zhaochenyang20 commented:
@llmforever hello. Sorry, I haven't noticed this before. Do you still need to fix this? Actually, we have a unit test for this in `test/srt/models/test_embedding_models.py`. The logits here is indeed closed.

Also, I don't understand what did you mean by "perform not so well". Could you provide your running snifts and your serving command for SGLang.

And, does e5-mistral also have this problem? Or only get?

thomZ1 commented:
![image](https://github.com/user-attachments/assets/6c3ff962-9e62-458b-841a-3e390349724a)
The same problem, I tried using the SGLang OpenAI API and SentenceTransformer with the same prompt, but the output embeddings were different.

zhaochenyang20 commented:
Yeah. The embedding could be different due to a lot of reasons. @llmforever 

You can check this unit test: https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_embedding_models.py

We set a tolerance value for the embedding difference.

Also, please try the e5-mistral model and give us the embedding difference. 

https://huggingface.co/intfloat/e5-mistral-7b-instruct

@Ying1123 Do you think the difference provided is tolerable?

llmforever commented:
> Yeah. The embedding could be different due to a lot of reasons. @llmforever
> 
> You can check this unit test: https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_embedding_models.py
> 
> We set a tolerance value for the embedding difference.
> 
> Also, please try the e5-mistral model and give us the embedding difference.
> 
> https://huggingface.co/intfloat/e5-mistral-7b-instruct
> 
> @Ying1123 Do you think the difference provided is tolerable?

I test about 10 cases，**each accuracy drop from 80% to less than 10%**，i think the difference is not tolerable，but the result of the e5-mistral-7b-instruct model is the same，can you please help me look that? Here is the code i use to generate the embedding：

**for transformer：**

import torch
import torch.nn.functional as F

from torch import Tensor
from transformers import AutoTokenizer, AutoModel

input_texts = ['hello']
tokenizer = AutoTokenizer.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)
model = AutoModel.from_pretrained('Alibaba-NLP/gte-Qwen2-7B-instruct', trust_remote_code=True)

max_length = 8192

batch_dict = tokenizer(input_texts, max_length=max_length, padding=True, truncation=True, return_tensors='pt')
outputs = model(**batch_dict)
embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

# normalize embeddings
embeddings = F.normalize(embeddings, p=2, dim=1)

**for sglang:**
import openai
import time
client = openai.Client(
    base_url="http://localhost:30000/v1", api_key="EMPTY")

input_texts = ['hello']

queres = client.embeddings.create(
    model="default",
    input=quelist,
)
embeddings  = torch.tensor(response.data[0].embedding)
